crowbar_framework/lib/crowbar/lock/shared_non_blocking.rb

```diff
@@ -0,0 +1,122 @@
+#
+# Copyright 2011-2013, Dell
+# Copyright 2013-2015, SUSE LINUX GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+module Crowbar
+  class Lock::SharedNonBlocking < Lock
+    # This class implements a shared (i.e. non-exclusive) lock, by
+    # default on a remote node.  It is used to allow one or more
+    # concurrent proposal applications to temporarily pause any
+    # intervallic runs triggered by the chef-client daemon, to prevent
+    # them from interfering with the orchestration of the proposals
+    # being applied:
+    #
+    #   https://bugzilla.suse.com/show_bug.cgi?id=857375
+    #
+    # It has the same semantics as a shared read/write lock, except
+    # that instead of readers reading, we have threads applying
+    # proposals, and instead of a single writer writing, we have the
+    # chef-client daemon triggering a chef-client run.
+
+    attr_reader :node, :owner, :reason, :executor
+
+    def initialize(options = {})
+      super
+      %w(node owner reason).each do |option|
+        unless options[option.to_sym]
+          raise "#{self.class} can't be instantiated without " + \
+                ":#{option} option (#{options.inspect})"
+        end
+        instance_variable_set "@#{option}", options[option.to_sym]
+      end
+
+      default_executor = lambda { |node, cmd|
+        ret = RemoteNode.ssh_cmd(node, cmd)
+        raise $? unless ret
+        ret
+      }
+      @executor = options.fetch :executor, default_executor
+
+      # To implement shared locking semantics, we keep track of all
+      # "readers" via a 'foo.lock.d/' directory which contains one
+      # file per "reader":
+      @readers_dir = path + '.d'
```

Comment:
Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.